merlin-server/sql/merlin/tables/activity_directive.sql

```diff
@@ -68,6 +80,23 @@ $$;
 comment on procedure plan_locked_exception(plan_id integer) is e''
   'Verify that the plan corresponding to the activity being updated is unlocked, throwing an exception if not.';
 
+create procedure nothing_anchored_before_delete(activity_id integer)
```

Comment:
I wonder if we can get this "restrict deletion" behavior by leveraging the foreign key, up on line 30. I think if we say `on delete restrict`, we'll get the same behavior as with this trigger

Reply:
Looking into this more, it looks like we can make the order of operations within a transaction a bit more flexible with [`on delete no action`](https://www.postgresql.org/docs/current/ddl-constraints.html#DDL-CONSTRAINTS-FK) and [`deferrable initially deferred`](https://www.postgresql.org/docs/current/sql-set-constraints.html) and delay checking if the foreign key is satisfied until the COMMIT step of the transaction. Thoughts?

Reply:
Thanks for sharing those docs - it's good to know that we have options. I like your thought about the order of delete statements - it is true that deferring the check could allow deletes to occur in multiple orders - but users shouldn't be doing bulk deletes in separate statements. The most idiomatic way to do a bulk delete is the following:

```sql
delete from activity_directive where plan_id = 1 and activity_id in (1, 2, 3, 4, 5);
```

(or the graphql equivalent, using `_in: [1, 2, 3, 4, 5]`)

This deletes the whole set atomically - so there is no observable state that could violate our constraints.

So, let’s go for “on delete restrict” here until we encounter a compelling reason to defer the check.

Reply:
Good to note, but I was thinking about the case where we want to delete an anchor and reanchor the thing anchored to it.

ie B is anchored to A and we want to delete A. With `on delete restrict`, the transaction must be
```sql
update activity_directive
set anchor_id = null
where (id, plan_id) = (B.id, B.plan_id);

delete from activity_directive
where (id, plan_id) = (A.id, A.plan_id);
```

But with the deferral, we could instead chose to reverse the order of those statements if we want.

Reply:
Follow up on this: while we do get the same behavior using `on delete restrict` as we do with the trigger, I think the error message gotten from `on delete restrict` is more difficult to understand.

on delete restrict: `[23503] ERROR: update or delete on table "activity_directive" violates foreign key constraint "anchor_in_plan" on table "activity_directive" Detail: Key (id, plan_id)=(1, 1) is still referenced from table "activity_directive".`

Trigger: `[P0001] ERROR: Cannot delete. Other activity directives are anchored to Activity Directive 1.`

Reply:
That's a good point :+1: I think that's a broader question than this ticket though - all of our constraint violation error messages pass the postgres error messages to the API client.

Looking around, hasura has an open ticket regarding abilities to customize error messages, so at least we know other people have had this thought as well https://github.com/hasura/graphql-engine/issues/3658